### PR TITLE
Update applications info stored in the filestore when deleting apps

### LINKED ISSUE #4941: Metrics `insight_application_total ` isn't zero after deleting the applications in a project
**What happened**:

After deleting the applications in a project, the metrics `insight_application_total` isn't zero.

>  insight_application_total | gauge | Number of applications currently controlled by control plane.

https://pipecd.dev/docs-v0.47.x/user-guide/metrics/#control-plane-metrics

**What you expected to happen**:

Got zero value or nothing after deleting the applications in a project.

**How to reproduce it**:

As an environment, I used a Control Plane in the kind cluster and piped on my local machine.

Create a project called `test` and two apps for the project.
Then, get the metrics `insight_application_total` before and after deleting apps from `/metrics` in the ops server. (port-forwarded to the pod)

### Before

**Applications**
_CS__PipeCD上でApplicationをすべて削除したのに、opsのメトリクス_`insight_application_total`_が_0にならない_·_Issue__100_·_ca-dp_ffjlabo-dev

**Metrics**
```
% curl http://localhost:9085/metrics | grep insight_application_total
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --100  6888    0  6888    0     0   156k      0 --:--:-- --:--:-- --:--:--  168k
# HELP insight_application_total Number of applications currently controlled by control plane
# TYPE insight_application_total gauge
insight_application_total{app_kind="KUBERNETES",pipecd_component="ops",project="test"} 2
```

### After deleting the apps

**Applications**

PipeCD

**Metrics**

```
% curl http://localhost:9085/metrics | grep insight_application_total
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --100  6908    0  6908    0     0   140k      0 --:--:-- --:--:-- --:--:--  149k
# HELP insight_application_total Number of applications currently controlled by control plane
# TYPE insight_application_total gauge
insight_application_total{app_kind="KUBERNETES",pipecd_component="ops",project="test"} 1
```

**Environment**:
- `piped` version:
- `control-plane` version:
- Others:

### PR DESCRIPTION
**Pull request by ffjlabo**
**What this PR does / why we need it**:

This PR is that the ops in the Cotrol Plane don't show `insight_applications_total` for the project deleted all apps.

We can get the metrics from the ops component in the Control Plane by calling `/metrics`.
`insight_applications_total` is the metric created by the flow below.

insight-metrics-total-before

Currently, the insight collector doesn't update application info in the filestore for the project which is deleted all apps.
This is because the only not deleted application info is updated after deleting all apps in the project.

So I fixed to update the app info as an empty array after deleting all apps in the project.

**Fixes**
- get all status apps and check them later to init `appsByProject` to store the app info as an empty array [feae904](https://github.com/pipe-cd/pipecd/pull/4953/commits/feae904e8b639251d01fec05349f0344e15dc100)

**Which issue(s) this PR fixes**:

Fixes #4941

**Does this PR introduce a user-facing change?**:

- **How are users affected by this change**:
- **Is this breaking change**:
- **How to migrate (if breaking change)**:

### CODE REVIEW

**Comment by ffjlabo** [comment]
**checking the behavior**

not to show the metric after deleting the apps in the project `test`.

before deleting apps
```
% curl http://localhost:9085/metrics | grep insight_application_total
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --100  6910    0  6910    0     0   151k      0 --:--:-- --:--:-- --:--:--  173k
# HELP insight_application_total Number of applications currently controlled by control plane
# TYPE insight_application_total gauge
insight_application_total{app_kind="KUBERNETES",pipecd_component="ops",project="test"} 2
```

after deleting apps
```
% curl http://localhost:9085/metrics | grep insight_application_total
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --100  6691    0  6691    0     0   154k      0 --:--:-- --:--:-- --:--:--  171k
```

**Review by t-kikuc — DISMISSED** [review]
Thanks, I understood!

File: pkg/app/ops/insightcollector/application_collector.go [hunk 1]
```diff
@@ -88,13 +97,7 @@ func (c *applicationDataCollector) listApplications(ctx context.Context) ([]*mod
 
 	for {
 		apps, next, err := c.lister.List(ctx, datastore.ListOptions{
-			Filters: []datastore.ListFilter{
-				{
-					Field:    "Deleted",
-					Operator: datastore.OperatorEqual,
-					Value:    false,
-				},
-			},
```
Comment (khanhtc1202) [inline pkg/app/ops/insightcollector/application_collector.go:97]: Sorry, I don't understand why do we have to remove this filter value from the query, could you explain?

File: pkg/app/ops/insightcollector/application_collector.go [hunk 2]
```diff
@@ -88,13 +97,7 @@ func (c *applicationDataCollector) listApplications(ctx context.Context) ([]*mod
...
+			Filters: []datastore.ListFilter{},
```
Comment (khanhtc1202) [inline pkg/app/ops/insightcollector/application_collector.go:100]: Nice catch 👍 I think we should add a comment to the head of this `listApplications` function (L93) to explain its behavior. Even if it looks obvious, it helps to mention that we expect this function to return values that way. Something like

```
// listApplications queries the datastore and gets all projects' applications (including disabled/deleted apps)
// to ensure projects with all applications disabled/deleted are listed as well.
```

File: pkg/app/ops/insightcollector/application_collector.go [hunk 3]
```diff
@@ -59,6 +59,15 @@ func (c *applicationDataCollector) Execute(ctx context.Context) {
...
 			app     = insight.BuildApplicationData(a)
 		)
+
+		if _, ok := appsByProject[project]; !ok {
+			appsByProject[project] = make([]*insight.ApplicationData, 0)
+		}
+
+		if app.Status == model.ApplicationActiveStatus_DELETED.String() {
+			continue
+		}
```
Comment (khanhtc1202) [inline pkg/app/ops/insightcollector/application_collector.go:69]:
```suggestion
		// Do not count the deleted application.
		if app.Status == model.ApplicationActiveStatus_DELETED.String() {
			continue
		}
```
nits

**Comment by khanhtc1202** [comment]
Nice work @ffjlabo I just left some nits 🙏

File: pkg/app/ops/insightcollector/application_collector.go (hunk 3 above)
Comment (ffjlabo) [reply pkg/app/ops/insightcollector/application_collector.go]: Thanks, added it! [c657a23](https://github.com/pipe-cd/pipecd/pull/4953/commits/c657a2370163be6cfc24b2448b1069ec4ba67d51)

File: pkg/app/ops/insightcollector/application_collector.go (hunk 2 above)
Comment (ffjlabo) [reply pkg/app/ops/insightcollector/application_collector.go:103]: Thanks, added it! [dd971aa](https://github.com/pipe-cd/pipecd/pull/4953/commits/dd971aaa7b8810bacd9470cfd863cce6b775551d)

**Comment by ffjlabo** [comment]
@khanhtc1202 thanks for the comment, I fixed them🙏

**Review by khanhtc1202 — APPROVED** [review]
nice catch, thanks 🚀

**Pull request merged**